### PR TITLE
Mldb 1717 query route supports body

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: server/mldb_server.cc [hunk 1]
```diff
@@ -186,11 +186,15 @@ initRoutes()
 
    // MLDB-1380 - make sure that the CPU support the minimal instruction sets
     if (supportsSystemRequirements()) {
+        const auto queryStringDef = "The SQL query string. Must be defined "
```
Comment (mailletf) [inline server/mldb_server.cc]: Pour éviter la confusion entre query string et query string, je dirais: The string representing the SQL query. Must be ...

File: server/mldb_server.cc [hunk 2]
```diff
@@ -186,11 +186,15 @@ initRoutes()
...
+                                    "either as a query string parameter or "
+                                    "in the body.";
```
Comment (mailletf) [inline server/mldb_server.cc]: or as JSON in the body ?

File: server/mldb_server.cc [hunk 3]
```diff
@@ -234,23 +241,39 @@ initRoutes()
...
-             bool sortColumns) const
+             bool sortColumns,
+             const Utf8String & bQuery
+             ) const
 {
-    auto stm = SelectStatement::parse(query.rawString());
+    if (qsQuery == "" && bQuery == "") {
+        throw ML::Exception("q is currently undefined. It must be defined "
+                            "either as a query string parameter or as a key "
+                            "in body data.");
```
Comment (mailletf) [inline server/mldb_server.cc]: data -> payload? (pas sûr)

File: server/mldb_server.cc [hunk 4]
```diff
@@ -234,23 +241,39 @@ initRoutes()
...
+    }
+
+    if (qsQuery != "" && bQuery != "") {
+        throw ML::Exception("q is currently defined twice. It must be defined "
+                            "either as a query string parameter or as a key "
+                            "in body data.");
```
Comment (mailletf) [inline server/mldb_server.cc]: same

File: server/mldb_server.cc [hunk 5]
```diff
@@ -186,11 +186,15 @@ initRoutes()
 
    // MLDB-1380 - make sure that the CPU support the minimal instruction sets
     if (supportsSystemRequirements()) {
+        const auto queryStringDef = "The string representing the SQL query. "
+                                    "Must be defined either as a query string "
+                                    "parameter or as JSON the body.";
```
Comment (mailletf) [inline server/mldb_server.cc]: JSON the body -> the JSON body

File: server/mldb_server.h [hunk 6]
```diff
@@ -115,13 +115,14 @@ struct MldbServer: public ServicePeer, public EventRecorder {
...
-    void runHttpQuery(const Utf8String& query,
+    void runHttpQuery(const Utf8String& qsQuery,
                       RestConnection & connection,
                       const std::string & format,
                       bool createHeaders,
                       bool rowNames,
                       bool rowHashes,
-                      bool sortColumns) const;
+                      bool sortColumns,
+                      const Utf8String & fmlh) const;
```
Comment (mailletf) [inline server/mldb_server.h]: nom de la variable fmlh? :P

**Comment by mailletf** [comment]
+1 quand les commentaires sont adressés

**Comment by guyd** [comment]
It is common to use a POST in such a case.  It is uncommon for a REST server to parse the body of a GET request.

**Comment by mailletf** [comment]
Uncommon but not unrestful. We went in that direction because POST is for ressource creation, while GET is to retrieve information.

**Pull request merged**